engine/src/main/java/com/google/android/fhir/sync/Sync.kt

```diff
@@ -147,14 +248,95 @@ object Sync {
         it.backoffCriteria.timeUnit,
       )
       periodicWorkRequestBuilder.setInputData(
-        Data.Builder().putInt(MAX_RETRIES_ALLOWED, it.maxRetries).build(),
+        Data.Builder()
+          .putInt(MAX_RETRIES_ALLOWED, it.maxRetries)
+          .putString(SYNC_STATUS_PREFERENCES_DATASTORE_KEY, uniqueWorkName)
+          .build(),
       )
     }
     return periodicWorkRequestBuilder.build()
   }
 
   /** Gets the timestamp of the last sync job. */
   fun getLastSyncTimestamp(context: Context): OffsetDateTime? {
-    return DatastoreUtil(context).readLastSyncTimestamp()
+    return FhirEngineProvider.getFhirDataStore(context).readLastSyncTimestamp()
   }
+
+  private fun createSyncState(
+    workInfoState: WorkInfo.State,
+    currentSyncJobStatus: SyncJobStatus,
+    terminalSyncJobStatus: SyncJobStatus,
+  ): SyncState {
+    return when (workInfoState) {
+      ENQUEUED -> {
+        Enqueued
+      }
+      RUNNING -> {
+        Running(currentSyncJobStatus)
+      }
+      SUCCEEDED -> {
+        Succeeded(terminalSyncJobStatus)
+      }
+      FAILED -> {
+        Failed(terminalSyncJobStatus)
+      }
+      CANCELLED -> {
+        Cancelled
+      }
+      else -> {
+        Unknown
+      }
+    }
+  }
+
+  /**
+   * Maps the [lastSyncJobStatus] to a specific [Result] based on the provided status.
+   *
+   * @param lastSyncJobStatus The last synchronization job status of type [SyncJobStatus].
+   * @return The mapped [Result] based on the provided [lastSyncJobStatus]:
+   * - [Result.Succeeded] with the timestamp if the last job status is [SyncJobStatus.Finished].
+   * - [Result.Failed] with exceptions and timestamp if the last job status is
+   *   [SyncJobStatus.Failed].
+   * - `null` if the last job status is neither Finished nor Failed.
+   */
+  private fun mapSyncJobStatusToResult(
+    lastSyncJobStatus: SyncJobStatus,
+  ) =
+    when (lastSyncJobStatus) {
+      is SyncJobStatus.Finished -> Result.Succeeded(lastSyncJobStatus.timestamp)
+      is SyncJobStatus.Failed ->
+        Result.Failed(lastSyncJobStatus.exceptions, lastSyncJobStatus.timestamp)
+      else -> Result.Unknown
+    }
```

Comment:
cleaned up to take care of case where there is no lastSyncJobStatus
```suggestion
  private fun mapSyncJobStatusToResult(
    lastSyncJobStatus: SyncJobStatus?,
  ) =
    lastSyncJobStatus?.let {
      when (it) {
        is SyncJobStatus.Finished -> Result.Succeeded(it.timestamp)
        is SyncJobStatus.Failed -> Result.Failed(it.exceptions, lastSyncJobStatus.timestamp)
        else -> error("Can't have")
      }
    }
```

Reply:
done.